scalameta/parsers/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala

```diff
@@ -1089,7 +1089,10 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
       simpleTypeRest(autoPos(token match {
         case LeftParen()  => autoPos(makeTupleType(inParens(types())))
         case Underscore() => next(); atPos(in.prevTokenPos, auto)(Type.Placeholder(typeBounds()))
-        case _       =>
+        case _ if token.is[Literal] =>
+          if (dialect.allowLiteralTypes) literal()
+          else throw new UnsupportedOperationException(s"$dialect doesn't support literal types")
```

Comment:
Also, the pattern match can be written a little bit more concisely via `case Literal() => ...`.